velox/exec/Spiller.cpp

```diff
@@ -341,73 +368,59 @@ bool Spiller::fillSpillRuns(
+      const auto partition = (type_ == Type::kOrderBy)
+          ? 0
+          : bits_.partition(hashes[i], state_.maxPartitions());
+      VELOX_DCHECK_GE(partition, 0);
+      if (FOLLY_UNLIKELY(
+              rowsFromNonSpillingPartitions != nullptr &&
+              !state_.isPartitionSpilled(partition))) {
+        rowsFromNonSpillingPartitions->push_back(rows[i]);
         continue;
       }
       spillRuns_[partition].rows.push_back(rows[i]);
       spillRuns_[partition].numBytes += container_.rowSize(rows[i]);
     }
-    // The final phase goes through the whole container and makes runs for all
-    // non-empty spilling partitions.
-    if (final && numRows) {
-      continue;
-    }
-    bool anyStarted = false;
-    for (auto i = 0; i < spillRuns_.size(); ++i) {
-      auto& run = spillRuns_[i];
-      if (!run.rows.empty() && (run.numBytes > targetSize || final)) {
-        pendingSpillPartitions_.insert(i);
-        anyStarted = true;
-      }
-    }
-    if (final) {
-      return true;
-    }
     if (numRows == 0) {
-      if (numConsidered == container_.numRows()) {
-        // If done full sweep but no spill started yet, start enough partitions
-        // to cover the ask.
-        std::vector<int32_t> indices(spillRuns_.size());
-        std::iota(indices.begin(), indices.end(), 0);
-        std::sort(
-            indices.begin(), indices.end(), [&](int32_t left, int32_t right) {
-              return spillRuns_[left].numBytes > spillRuns_[right].numBytes;
-            });
-        int64_t started = 0;
-        for (auto i : indices) {
-          if (spillRuns_[i].numBytes > 0) {
-            pendingSpillPartitions_.insert(i);
-            started += spillRuns_[i].numBytes;
-          }
-          if (started > targetSize) {
-            break;
-          }
-        }
-        clearNonSpillingRuns();
-        return started == 0;
-      }
-      clearNonSpillingRuns();
-      return true;
-    }
-    if (anyStarted) {
-      clearNonSpillingRuns();
-      return false;
+      break;
     }
   }
 }
 
 void Spiller::clearNonSpillingRuns() {
-  for (auto i = 0; i < spillRuns_.size(); ++i) {
-    if (pendingSpillPartitions_.find(i) == pendingSpillPartitions_.end()) {
-      spillRuns_[i].clear();
+  for (auto partition = 0; partition < spillRuns_.size(); ++partition) {
+    if (pendingSpillPartitions_.count(partition) == 0) {
+      spillRuns_[partition].clear();
     }
   }
 }
 
+std::string Spiller::toString() const {
+  return fmt::format(
+      "Type{}\tRowType:{}\tNum Partitions:{}\tFinalized:{}",
+      TypeName(type_),
+      rowType_->toString(),
+      state_.maxPartitions(),
+      spillFinalized_);
+}
+
+// static
+std::string Spiller::TypeName(Type type) {
+  switch (type) {
+    case Type::kOrderBy:
+      return "ORDER_BY";
+    case Type::kHashJoin:
+      return "HASH_JOIN";
+    case Type::kAggregate:
+      return "AGGREGATE";
+    default:
+      LOG(DFATAL) << "Unknown type: " << static_cast<int>(type);
```

Comment:
Why not just VELOX_FAIL or VELOX_UNREACHABLE?

Reply:
Do we have a fatal log only in debug mode? sth like VELOX_DFAIL? Thanks!